File: src/main/java/co/herod/adbwrapper/AdbDeviceActions.java
package co.herod.adbwrapper;

import co.herod.adbwrapper.model.AdbDevice;
import co.herod.adbwrapper.util.UiHierarchyHelper;
import io.reactivex.Completable;
import io.reactivex.disposables.Disposable;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.concurrent.TimeUnit;

@SuppressWarnings({"unused", "WeakerAccess"})
public class AdbDeviceActions {

    public static final int KEY_EVENT_HOME = 3;
    public static final int KEY_EVENT_BACK = 4;
    public static final int KEY_EVENT_POWER = 26;

    public static void pressHomeButton(@Nullable final AdbDevice adbDevice) {
        Adb.pressKeyBlocking(adbDevice, KEY_EVENT_HOME);
    }

    public static void pressBackButton(@Nullable final AdbDevice adbDevice) {
        Adb.pressKeyBlocking(adbDevice, KEY_EVENT_BACK);
    }

    public static void pressPowerButton(@Nullable final AdbDevice adbDevice) {
        Adb.pressKeyBlocking(adbDevice, KEY_EVENT_POWER);
    }

    public static void turnDeviceScreenOn(@NotNull final AdbDevice adbDevice) {
        while (!AdbDeviceProperties.isScreenOn(adbDevice)) pressPowerButton(adbDevice);
    }

    public static void turnDeviceScreenOff(@NotNull final AdbDevice adbDevice) {
        while (AdbDeviceProperties.isScreenOn(adbDevice)) pressPowerButton(adbDevice);
    }

    public static void tapCentre(final AdbDevice connectedAdbDevice, final Integer[] c) {

        Completable.fromObservable(ProcessHelper.observableProcess(AdbProcesses.tap(connectedAdbDevice, UiHierarchyHelper.centreX(c), UiHierarchyHelper.centreY(c))))
                .blockingAwait(5, TimeUnit.SECONDS);
    }

    @NotNull
    public static Disposable tapUiNode(final AdbDevice connectedAdbDevice, @NotNull final String s) {
        return UiHierarchyHelper.extractBoundsInts(s).subscribe(c -> tapCentre(connectedAdbDevice, c));
    }

    public static boolean tapCoords(final AdbDevice connectedAdbDevice, final int x, final int y) {

        return Completable.fromObservable(ProcessHelper.observableProcess(AdbProcesses.tap(connectedAdbDevice, x, y)))
                .blockingAwait(5, TimeUnit.SECONDS);
    }
}

File: src/main/java/co/herod/adbwrapper/AdbDeviceActions.kt
package co.herod.adbwrapper

import co.herod.adbwrapper.model.AdbDevice
import co.herod.adbwrapper.model.AdbUiNode
import co.herod.adbwrapper.util.UiHierarchyHelper
import io.reactivex.Completable
import java.util.concurrent.TimeUnit

object AdbDeviceActions {

    private const val KEY_EVENT_HOME = 3
    private const val KEY_EVENT_BACK = 4
    private const val KEY_EVENT_POWER = 26
    private const val KEY_EVENT_BACKSPACE = 67

    fun AdbDevice?.pressHomeButton() {
        Adb.pressKeyBlocking(this, KEY_EVENT_HOME)
    }

    fun AdbDevice?.pressBackButton() {
        Adb.pressKeyBlocking(this, KEY_EVENT_BACK)
    }

    fun AdbDevice?.pressPowerButton() {
        Adb.pressKeyBlocking(this, KEY_EVENT_POWER)
    }

    fun AdbDevice?.pressBackspaceButton() {
        Adb.pressKeyBlocking(this, KEY_EVENT_BACKSPACE)
    }

    fun turnDeviceScreenOn(adbDevice: AdbDevice?) {
        while (!AdbDeviceProperties.isScreenOn(adbDevice)) adbDevice.pressPowerButton()
    }

    fun turnDeviceScreenOff(adbDevice: AdbDevice?) {
        while (AdbDeviceProperties.isScreenOn(adbDevice)) adbDevice.pressPowerButton()
    }

    fun AdbDevice.tapCentre(adbUiNode: AdbUiNode) = this.tapCentre(adbUiNode.bounds!!)

    private fun AdbDevice.tapCentre(c: Array<Int>) = ProcessHelper.blocking(AdbProcesses.tap(this,
            UiHierarchyHelper.centreX(c),
            UiHierarchyHelper.centreY(c)))

    fun AdbDevice.tapUiNode(s: AdbUiNode) {
        UiHierarchyHelper.extractBoundsInts(s)?.subscribe { tapCentre(it) }
    }

    fun tapCoords(adbDevice: AdbDevice, x: Int, y: Int): Boolean =
            Completable.fromObservable(ProcessHelper.observableProcess(AdbProcesses.tap(adbDevice, x, y)))
                    .blockingAwait(5, TimeUnit.SECONDS)
}
